### ISSUE TITLE
Poor performance when log size increases

### ISSUE THREAD

**Issue by armunro** [armunro/git-tfs#1]
**Background**
Console output and logging are both handled by the System.Diagostics.Trace type that offers a simple static method for logging diagnostic info to a configurable set of listeners.

**Problem**
For unknown reasons, The performance of Trace.* calls correlates with the size of the log file in C:\Users\....\AppData\Local\git-tfs\git-tfs_log.txt. As the log file becomes larger, every Trace.* call begins to take longer to complete. At the worst point, a log file of 171MB was adding 700ms per console line. Typically user logs will not get this large, but in this case, a scheduled task was configured to fetch every 5 minutes . This is either a fundamental flaw in System.Diagnostics.Trace or NLog (more likely). Deleting the log file eliminated any console delay and improved performance beyond baseline.

**Goals**
Logging and console output concerns are a very small portion of the code base and may not require enterprise level logging. A simple output writer that is able to write to console and log files will be sufficient and will remove the dependency on NLog,

**Comment by mpmel** [comment]
is this completely fixed by [e9444e2b091c99d90907f9cfb9ddbe2e4ae2a8dc](https://github.com/armunro/git-tfs/commit/e9444e2b091c99d90907f9cfb9ddbe2e4ae2a8dc)?

I'm working on speeding up git-tfs in my own fork as the main repo seems not to not merge many pull requests and there are quite a few that look useful/improve performance.

I added a pull request there yesterday [#1290](https://github.com/git-tfs/git-tfs/pull/1290) and am just looking for some more things I can merge into my own to make it faster and more resilient as we have a large a migration from TFS to Git happening soon and there are more than 100k checkins across a lot of different branches that need to be brought in

**Comment by armunro** [comment]
@mpmel - Apologies for the delay. I was right in the middle of a complete re-organization of the entire codebase when I got pulled on to a high priority project. I think this commit is safe to take but the refactor as a whole needs more testing before I can feel comfortable. That being said, there should be no functional changes aside from logging that were made in this commit.

As a quick-fix, deleting the log file intermittently throughout the fetch process will keep the log file small which will speed up the fetch as a whole. This problem only becomes apparent when the log file gets large.

**Comment by mpmel** [comment]
Hi, no problem. I was just wondering if you had any metrics on the speed-up. I took your log changes and also my pull request and one other performance one and it significantly sped up my fetch time. 

I ended up able to pull in my ~18,000 commits across 6 or 7 branches in about 3 hours. I don't have exact numbers but I think previously it was 6-8 hours.

I think I'm currently left mostly with just the time it takes to pull down all of the files and do the TFS operations.

Also, oddly enough, I noticed that sometimes the Trace lines get hung. The process just pauses for a while and if I click the Console and press Esc, it will kick off again. I'm not sure if this is some kind of bug or know thing with the Trace operation, but that may also be what's causing it to take a little longer sometimes.

**Comment by armunro** [comment]
Awesome. Three hours for that many commits is a pretty significant improvement over what I've seen.

Interesting note on the Trace lines hanging. I haven't noticed this but I will be keeping an eye out for it. Eventually, I will be replacing the Trace lines with a custom singleton that shouldn't come with any of this overhead. 

I should also mention that the way I am saving to the log file is not ideal yet. NLog and Log4Net do background IO to prevent logging performance from impacting the main thread. In reality, Git-Tfs probably doesn't produce enough log entries for this to be a factor but it should still be fixed.

**Comment by mpmel** [comment]
Yes, the pull request 1290 that I made there was to cache the workspaces during the fetch. It was completely setting up and tearing down the entire workspace for every commit individually. On my system that was adding 600-800ms per commit in debug.

As of now, I don't really know of anything else simple that would speed it up. Most of the time left is just in the Get operation from TFS. Potentially coulld multi-thread the get operation into multiple separate workspaces to make it a bit faster, but I'm mostly preparing for a one-off migration to Git, so I don't know that I will go that far.
